plugins/alembic/module/vtkF3DAlembicReader.cxx

```diff
@@ -20,51 +22,338 @@
+    const Alembic::AbcGeom::IPolyMeshSchema& schema = pmesh.getSchema();
+    if (schema.getNumSamples() > 0)
     {
-      pmesh.getSchema().get(samp);
+
+      schema.get(samp);
 
       Alembic::AbcGeom::P3fArraySamplePtr positions = samp.getPositions();
-      Alembic::AbcGeom::Int32ArraySamplePtr indices = samp.getFaceIndices();
-      Alembic::AbcGeom::Int32ArraySamplePtr counts = samp.getFaceCounts();
+      Alembic::AbcGeom::Int32ArraySamplePtr face_positionIndices = samp.getFaceIndices();
+      Alembic::AbcGeom::Int32ArraySamplePtr faceVertexCounts = samp.getFaceCounts();
 
-      points->SetNumberOfPoints(positions->size());
+      this->SetupIndicesStorage(faceVertexCounts, originalData.Indices);
 
-      for (size_t i = 0; i < positions->size(); i++)
+      // Positions
       {
-        points->SetPoint(i, positions->get()[i].x, positions->get()[i].y, positions->get()[i].z);
-      }
+        V3fContainer pV3F;
+        for (size_t pIndex = 0; pIndex < positions->size(); pIndex++)
+        {
+          pV3F.emplace_back(
+            positions->get()[pIndex].x, positions->get()[pIndex].y, positions->get()[pIndex].z);
+        }
+        originalData.Attributes.insert(AttributesContainer::value_type("P", pV3F));
 
-      vtkNew<vtkIdTypeArray> offsets;
-      vtkNew<vtkIdTypeArray> connectivity;
-      offsets->SetNumberOfTuples(counts->size() + 1);
-      connectivity->SetNumberOfTuples(indices->size());
+        this->UpdateIndices<Alembic::AbcGeom::Int32ArraySamplePtr>(
+          face_positionIndices, pIndicesOffset, originalData.Indices);
+      }
 
-      offsets->SetTypedComponent(0, 0, 0);
-      for (size_t i = 0; i < counts->size(); i++)
+      // Texture coordinate
+      Alembic::AbcGeom::IV2fGeomParam uvsParam = schema.getUVsParam();
+      if (uvsParam.valid())
       {
-        offsets->SetTypedComponent(i + 1, 0, offsets->GetTypedComponent(i, 0) + counts->get()[i]);
+        Alembic::AbcGeom::IV2fGeomParam::Sample uvValue = uvsParam.getIndexedValue();
+        if (uvValue.valid())
+        {
+          V3fContainer uvV3F;
+          Alembic::AbcGeom::UInt32ArraySamplePtr uvIndices = uvValue.getIndices();
+          for (size_t index = 0; index < uvValue.getVals()->size(); ++index)
+          {
+            Alembic::AbcGeom::V2f uv = (*(uvValue.getVals()))[index];
+            uvV3F.emplace_back(uv[0], uv[1], 0);
+          }
+          originalData.Attributes.insert(AttributesContainer::value_type("uv", uvV3F));
+
+          if (uvsParam.getScope() == Alembic::AbcGeom::kFacevaryingScope)
+          {
+            originalData.uvFaceVarying = true;
+            this->UpdateIndices<Alembic::AbcGeom::UInt32ArraySamplePtr>(
+              uvIndices, uvIndicesOffset, originalData.Indices);
+          }
+          else
+          {
+            this->UpdateIndices<Alembic::AbcGeom::Int32ArraySamplePtr>(
+              face_positionIndices, uvIndicesOffset, originalData.Indices);
+          }
```

Comment:
do you have non-kFacevaryingScope data to test this with @nyue ?

Reply:
I believe the first zip I attached in the comments above may fit this test requirements. Let me know if it is another corner case I need to generate specific geometry file for.